packages/components/src/components/SquiggleViewer/ViewerProvider.tsx

```diff
@@ -258,6 +347,48 @@ export const ViewerProvider: FC<
         case "UNREGISTER_ITEM_HANDLE":
           delete itemHandlesStoreRef.current[pathAsString(action.payload.path)];
           return;
+        case "CALCULATOR_NEW": {
+          const ref =
+            settingsStoreRef.current[pathAsString(action.payload.path)];
+          settingsStoreRef.current[pathAsString(action.payload.path)] = {
+            ...ref,
+            calculator: initialState(action.payload.calculator),
+          };
+          return;
+        }
+        case "CALCULATOR_SET_FIELD_CODE": {
+          changeCalculator(action.payload.path, (state) => {
+            const modifyField = (name: string, newField: FieldValue) => {
+              const newFields = { ...state.fields, [name]: newField };
+              return { ...state, fields: newFields };
+            };
+            const { name, code } = action.payload;
+            const field = state.fields[name];
+            const newValue = null;
+            const newField = { ...field, code, value: newValue };
+            return modifyField(name, newField);
+          });
+          return;
+        }
+        case "CALCULATOR_SET_FIELD_VALUE": {
+          changeCalculator(action.payload.path, (state) => {
+            const modifyField = (name: string, newField: FieldValue) => {
+              const newFields = { ...state.fields, [name]: newField };
+              return { ...state, fields: newFields };
+            };
+            const { name, value } = action.payload;
+            const field = state.fields[name];
+            const newField = { ...field, value };
+            return modifyField(name, newField);
+          });
+          return;
+        }
+        case "CALCULATOR_SET_FN_VALUE": {
+          changeCalculator(action.payload.path, (state) => {
+            return { ...state, fn: { value: action.payload.value } };
+          });
+          return;
+        }
```

Comment:
I'd like to clean most of this up into a different helper file. This huge ViewerProvider is getting gnarly, imo.

Reply:
I commented on managing the complexity above.

Now specifically on this part: I'd suggest a different approach where `ViewerProvider` shouldn't be aware of calculator state and treat it as a black box.

- add a separate `CalculatorProvider` in `Calculator` component; manage calculator state there with a native `useReducer` (grab the initial state value from `ViewerProvider`)
- have a single `CALCULATOR_SET` action in `ViewerProvider` component, with `CalculatorState` as a payload

Benefits:
- this dispatch function would never have to call `calculatorReducer`
- you won't have to call `forceUpdate` in calculator, since the native `useReducer` is reactive (it's limited to a single calculator, so performance won't be an issue)

Reply:
I'm confused. I could understand pulling out the coupling on ViewerProvider, but ``useReducer`` would manage storing state. Are you suggesting that we store state both in the ``useReducer`` call, and also update the ViewerProvider after each change?